File: P2/csci6511_p2_natavan_akhundova.py
import csv, sys, math, time
import matplotlib.pyplot as plt
import numpy as np

# additional variables
SLEEP_TIME = 0							# time between executing subsequent steps
INF = int(0x3f3f3f3f)					# infinity value
V = -1									# the number of vertices, default -1
K = -1									# the number of colors, default -1

# Main components for storage
vertices = {}							# stores vertice to color relationship
edges = {}								# stores vertice to vertice relationship

# A Function for Reading File Inputs
def read_input(filename):
	print("Reading file inputs...")
	time.sleep(SLEEP_TIME)

	global V, K # reference to global variables
	file = open(filename, 'r')
	lines = file.readlines()

	for line in lines:
		if line[0] == "#": # if a line is a comment
			continue
		line = line.strip()
		if K == -1:			# if the number of colors is not set
			K = int(line.split("=")[1].strip()) # set the number of colors
		else:
			line = line.split(",")
			from_vertice_id = int(line[0].strip())
			to_vertice_id = int(line[1].strip())

			# Vertice Assignment
			# default color for each vertex is -1
			vertices[from_vertice_id] = {"assigned":-1, "available":set([c for c in range(K)])}
			vertices[to_vertice_id] = {"assigned":-1, "available":set([c for c in range(K)])}

			# Edge Assignment
			# make edges adjacent in both ways since edges are undirected
			if from_vertice_id not in edges:
				edges[from_vertice_id] = []
			edges[from_vertice_id].append(to_vertice_id)

			if to_vertice_id not in edges:
				edges[to_vertice_id] = []
			edges[to_vertice_id].append(from_vertice_id)

	V = len(vertices) # set the number of vertices

	print("Done reading file inputs.")
	time.sleep(SLEEP_TIME)

# A Function to Check Whether The Current Assignment of Color Is Safe To Make:
#							(adjacent vertices should not be the same color)
def is_safe(adj_vertices, color):
	for adj_v in adj_vertices:
		# if adjacent vertex has the same color
		if vertices[adj_v]["assigned"] == color: 
			return False

	return True

# A Function to Return Least Constraining Values of Colors
def ordered_colors(available_colors, adj_vertices):
	colors_count = {k:0 for k in available_colors}
	for adj_v in adj_vertices:
		if vertices[adj_v]["assigned"] != -1:
			adj_color = vertices[adj_v]["assigned"]
			if adj_color in colors_count:
				colors_count[adj_color] += 1


	return sorted(colors_count, key=colors_count.get, reverse=True)

# A Recursive Helping Function for Coloring the Graph - Backtracking Search
def color_graph_recursive(v_id, keys):
	# if all vertices have been colored
	#if v_count == V:
	#	return True
	if v_id == -1:
		return True

	adj_vertices = edges[v_id] # get adjacent vertices of the vertex
	colors = ordered_colors(vertices[v_id]["available"], adj_vertices) # get least constraining minimum reamining variable

	# find a color in available colors for the vertex that is safe to assign
	for color in colors:
		if is_safe(adj_vertices, color):
			vertices[v_id]["assigned"] = color # assign the color

			next_vertex = -1
			min_remaining_value = INF
			for adj_v in adj_vertices:
				try:
					vertices[adj_v]["available"].remove(color)
				except:
					pass	
				finally:
					#print(v_id, adj_v, vertices[adj_v]["available"])
					if vertices[adj_v]["assigned"] == -1 and len(vertices[adj_v]["available"]) < min_remaining_value:
						min_remaining_value = len(vertices[adj_v]["available"])
						next_vertex = adj_v
			#print("next_vertex "+str(next_vertex))
			if color_graph_recursive(next_vertex, keys): # recursively color other vertices
				return True

			vertices[v_id]["assigned"] = -1 # remove assignment if it lead to no solution
			for adj_v in adj_vertices:
				vertices[adj_v]["available"].add(color)

	# no color has been assigned to the vertex v
	return False


# A Function for Finding a Solution to "Coloring the Graph" Problem
def color_graph():
	keys = list(vertices.keys())
	solution = color_graph_recursive(keys[0], keys) # start from the first vertex

	if solution == False:
		print("Solution does not exist")
	else:
		for v_id in keys:
			print("Color of %d is %d"%(v_id, vertices[v_id]["assigned"]))

# The Main Function
if __name__ == "__main__":
	# read inputs
	read_input(sys.argv[1])

	# call search function to color the map
	color_graph()

File: P2/Csci6511_p2_natavan_akhundova.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.*;

/**
 * Vertex --- a class for encapsulating vertex information.
 */
class Vertex{
    private int assigned = -1;          // a color assigned to the vertex, default -1
    private Set<Integer> available;     // colors that are assignable to the vertex (i.e domain)
    private Set<Integer> edges;         // adjacent vertices connected through edges

    /**
     * Constructor function for initializing variables
     * @param K - the number of colors given in the CSP
     */
    Vertex(int K){
        available = new HashSet<Integer>();
        for(int i = 0; i < K; i++){
            available.add(i);
        }
        edges = new HashSet<Integer>();
    }

    /**
     * Getter and setter functions for class variables
     */

    public int getAssigned(){
        return assigned;
    }

    public void setAssigned(int a){
        assigned = a;
    }

    public Set<Integer> getAvailable(){
        return available;
    }

    public void addAvailable(int a){
        available.add(a);
    }

    public void removeAvailable(int a){
        available.remove(a);
    }

    public Set<Integer> getEdges(){
        return edges;
    }

    public void addEdges(int a){
        edges.add(a);
    }
}
/**
 * Csci6511_p2_natavan_akhundova --- a program to solve CSP.
 */
public class Csci6511_p2_natavan_akhundova {
    static Map<Integer, Vertex> vertices;                   // storage for all vertices in CSP
    static Queue<int[]> arcs;                               // storage for all arcs/edges in CSP
    static Map<Integer, ArrayList<Integer>> removedColors;  // storage for removed colors during AC3 Checking
    static int K = -1;                                      // the number of colors given in the CSP

    /**
     * Checks if assignment of the color to the vertex is safe considering its neighbors
     * @param color - the assigned color to the vertex
     * @param adjacentVertices - a set of adjacent vertices of the vertex
     * @return boolean value
     */
    public static boolean isSafe(Integer color, Set<Integer> adjacentVertices){
        // if any of the adjacent vertices has the same color, it is not safe
        for (Integer adjVertexId : adjacentVertices){
            if (vertices.get(adjVertexId).getAssigned() == color) return false;
        }
        // otherwise, safe
        return true;
    }

    /**
     * Restores CSP to the state where colors were not deleted from domains
     * @param vertex - currently processing vertex object
     * @param adjacentVertices - a set of adjacent vertices of currently processing vertex
     * @param color - the assigned color to the vertex
     */
    public static void restorePreviousState(Vertex vertex, Set<Integer> adjacentVertices, int color){
        // remove assignment of the color to the vertex
        vertex.setAssigned(-1);

        // return the color to the adjacent vertices
        for (Integer adjVertexId : adjacentVertices) {
            vertices.get(adjVertexId).addAvailable(color);
        }

        // restore removed colors from AC-3 checking
        if (removedColors.size() > 0) {
            Set<Integer> keys = removedColors.keySet();
            for (int key : keys) {
                ArrayList<Integer> values = removedColors.get(key);
                for (int value: values) {
                    vertices.get(key).addAvailable(value);
                }
            }
        }
    }

    /**
     * Sorts colors in the domain according to constraints
     * @param availableColors - a set of colors in a domain of the vertex
     * @param adjacentVertices - a set of adjacent vertices of the vertex
     * @return an ArrayList
     */
    public static ArrayList<Integer> leastConstrainingColors(Set<Integer> availableColors,
                                                             Set<Integer> adjacentVertices){
        Map<Integer, Integer> colorCount = new HashMap<>();
        ArrayList<Integer> sortedColors = new ArrayList<>();

        // set occurrences of available colors to 0
        for (Integer color : availableColors){
            colorCount.put(color, 0);
        }

        // count and update occurrences (inversely) of colors of assigned adjacent vertices
        for (Integer adjVertexId : adjacentVertices){
            int adjacentColor = vertices.get(adjVertexId).getAssigned();
            if (adjacentColor != -1) {
                if (colorCount.containsKey(adjacentColor))
                    colorCount.put(adjacentColor, colorCount.get(adjacentColor) - 1); // decrease the count
            }
        }

        // sort by occurrence values in ascending order
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(colorCount.entrySet());
        list.sort(Map.Entry.comparingByValue());

        for (Map.Entry<Integer, Integer> entry : list) {
            sortedColors.add(entry.getKey());
        }
        return sortedColors;
    }

    /**
     * Checks arc consistency between two vertices
     * @param xId - id of the first vertex
     * @param yId - id of the second vertex
     * @return boolean
     */
    public static boolean AC(int xId, int yId){
        boolean removed = false;
        // for every value of x, there should exist allowable value of y in the arc
        Set<Integer> xDomain = vertices.get(xId).getAvailable(); // get available colors of the vertex
        Set<Integer> yDomain = vertices.get(yId).getAvailable(); // get available colors of the adjacent vertex

        int yElem = yDomain.iterator().next();
        if (yDomain.size() == 1 && xDomain.contains(yElem)) {
            vertices.get(xId).removeAvailable(yElem);
            if (!removedColors.containsKey(xId)) removedColors.put(xId, new ArrayList<>());
            removedColors.get(xId).add(yElem);
            removed = true;
        }

        return removed;
    }

    /**
     * Checks arc consistency of all CSP
     * @return boolean
     */
    public static boolean AC3() {
        Queue<int[]> arcsCopy = new LinkedList<>(arcs);
        removedColors = new HashMap<>();

        while (arcsCopy.size() != 0){
            int[] xy = arcsCopy.remove();
            int xId = xy[0];
            int yId = xy[1];

            if(AC(xId,yId))
                if(vertices.get(xId).getAvailable().size() == 0) return false;
        }

        return true;
    }

    /**
     * Colors the provided vertex and its adjacent vertices recursively checking FC and AC3
     * @param vertexId - id of the vertex to be colored
     * @return boolean
     */
    public static boolean colorGraphRecursive(Integer vertexId){
        // if there is no left vertex to color
        if (vertexId == -1)
            return true;

        Vertex vertex = vertices.get(vertexId); // get current vertex by Id
        Set<Integer> adjacentVertices = vertex.getEdges(); // get adjacent vertices
        ArrayList<Integer> colors = leastConstrainingColors(vertex.getAvailable(), adjacentVertices); // get least constraining colors to assign
        boolean failure = false;    // flag for failures of FC and AC3 checks within CSP

        // find a color in available colors for the vertex that is safe to assign
        for (Integer color : colors){
            if(isSafe(color, adjacentVertices)){
                vertex.setAssigned(color); // assign the current color to the current vertex

                int nextVertexId = -1;
                int minRemainingValue = Integer.MAX_VALUE;

                // ----------- Processing Adjacent Vertices -----------
                for (Integer adjVertexId : adjacentVertices){
                    Vertex adjVertex = vertices.get(adjVertexId); // get current adjacent vertex by Id
                    adjVertex.removeAvailable(color);

                    // Forward Checking: if any value has no values, terminate
                    if (adjVertex.getAvailable().size() == 0) {failure = true; break;}

                    // find the next vertex by the minimum remaining value
                    if (adjVertex.getAssigned() == -1 && adjVertex.getAvailable().size() < minRemainingValue){
                        minRemainingValue = adjVertex.getAvailable().size();
                        nextVertexId = adjVertexId;
                    }
                }
                // ----------- End Processing Adjacent Vertices -----------

                // 3-Consistency Checking
                if (!AC3()) failure = true;

                if (!failure && colorGraphRecursive(nextVertexId)) return true; // recursively color other vertices

                restorePreviousState(vertex, adjacentVertices, color); // restore the previous state if the color assignment lead to no solution
            }
        }

        // no color has been assigned to the vertex
        return false;
    }

    /**
     * Calls recursive function of coloring the graph and prints the result
     */
    public static void colorGraph(){
        boolean solution = false;
        Set<Integer> keys = vertices.keySet(); // get ids of all vertices

        // start coloring uncolored vertices
        for (Integer key : keys) {
            if(vertices.get(key).getAssigned() == -1) {
                solution = colorGraphRecursive(key);
                if (!solution) break; // if any vertex resulted in no assignment
            }
        }

        // printing the result
        if (!solution) System.out.println("Solution does not exist.");
        else{
            System.out.println("Solution exists:");
            for (Integer key : keys) {
                int color = vertices.get(key).getAssigned();
                // print id of vertices and their assigned colors colorful
                System.out.printf("\u001B[3%dmThe color of %d is %d.\u001B[3%dm\n",color+1, key, color, color+1);
            }
            System.out.println("\u001B[0m");
        }
    }

    /**
     * The Main Function; Handles reading an input file and starts the processing
     * @param args - should consist of an input file
     */
    public static void main(String args[])
    {
        // verify the input file
        if (args.length <= 0){
            System.out.println("Argument Not Found ✗\nExiting");
            System.exit(0);
        }

        String fileName = args[0];                  // get the name of the input file
        vertices = new HashMap<Integer, Vertex>(); // initialize storage for vertices
        arcs = new LinkedList<>();                 // initialize storage for arcs

        // ----------- Reading The File -----------
        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            System.out.println("Reading file inputs...");
            String line;
            int lineCount = 0;
            while ((line = br.readLine()) != null) {
                if(line.substring(0,1).equals("#"))
                    continue;
                if(K == -1) K = Integer.parseInt(line.split("=")[1].strip());
                else{
                    if(lineCount >= 5000) break; // limit the size of CSP

                    int fromVertice = Integer.parseInt(line.split(",")[0].strip());
                    int toVertice = Integer.parseInt(line.split(",")[1].strip());

                    // Add vertices to vertex collection
                    if (!vertices.containsKey(fromVertice)) vertices.put(fromVertice, new Vertex(K));
                    if (!vertices.containsKey(toVertice)) vertices.put(toVertice, new Vertex(K));

                    // Add undirected edges
                    vertices.get(fromVertice).addEdges(toVertice);
                    vertices.get(toVertice).addEdges(fromVertice);
                    lineCount++;
                }
            }
            System.out.println("Done reading file inputs ✓");

            // Store all edges of the graph
            Set<Integer> keys = vertices.keySet();
            for (int key : keys){
                Set<Integer> edges = vertices.get(key).getEdges();
                for (int edge : edges){
                    arcs.add(new int[]{key, edge});
                }
            }

            System.out.printf("The number of \n\t1)vertices: %d\n\t2)arcs: %d\n\t3)colors: %d\n",
                    vertices.size(),arcs.size(),K);
        }
        catch (Exception ex){
            ex.printStackTrace();
            System.out.println("File Name: Not Found ✗\nExiting");
            System.exit(0);
        }
        // ----------- End Reading The File -----------

        // ----------- Coloring The Graph -----------
        colorGraph();
    }
}
